### ISSUE TITLE
Loss function - How can I use L1 loss train the network?

### ISSUE THREAD

**Issue by foolhard** [tarashakhurana/4d-occ-forecasting#7]
Hello @tarashakhurana , 

I want to use L1 loss to train the network as your mentioned in your paper. But in this repo, L1 loss is not used. 

Could you provide the code for using L1 loss for training?

![image](https://github.com/tarashakhurana/4d-occ-forecasting/assets/13057970/21d5cdba-10a6-4bc4-a061-8d7419d689e0)

**Comment by tarashakhurana** [comment]
Hi, I'm not sure I understand your question. We _do_ use L1 loss for training, as specified [here](https://github.com/tarashakhurana/4d-occ-forecasting/blob/a7d394eef14151a561ca25b1685687dc2da804d1/train.sh#L7). One potential confusion to clarify is that we manually compute the gradient of the loss function with respect to the density grid (which is `grad_sigma` here). You can find that implemented [here](https://github.com/tarashakhurana/4d-occ-forecasting/blob/a7d394eef14151a561ca25b1685687dc2da804d1/lib/dvr/dvr.cu#L577). With the current code, only L1, L2 and AbsRel losses can be used to train the model. If you need to train with any other loss, I can provide a Differentiable Voxel Rendering layer in PyTorch that does not require manual gradient computation but has a huge memory footprint.

**Comment by foolhard** [comment]
Hi, thanks for the explanation. 
I want a Differential Voxel Rendering layer in PyTorch to try it more flexibly. Could you please share it?

**Comment by foolhard** [comment]
@tarashakhurana Referred to your previous repo [here](https://github.com/tarashakhurana/emergent-occ-forecasting/blob/main/model.py#L644), I implement a Differential Voxel Rendering layer and it works. 

However, I still want your code as a reference. Thanks a lot.

**Comment by tarashakhurana** [comment]
Thanks for the reminder! I have added it now: https://github.com/tarashakhurana/4d-occ-forecasting/tree/main#new-differentiable-voxel-rendering-implemented-as-a-layer-in-pytorch

**Comment by foolhard** [comment]
It worked. Thanks a lot.

**Comment by Orbis36** [comment]
I noticed the gradient from these 2 methods are quite different, the one in dvr.cu will actually converge faster, is there any explanation for that?
